### PR TITLE
[Mr. Minou's case] Add `offset` as fittable parameter, and allow user to specify their own databank

### PR DESCRIPTION
**Pull request by TranHuuNhatHuy**
This pull request is to implement 2 features, as suggested by Mr. @minouHub :
- Allow `offset` to be fitted. Previously, as `offset` and `slit` are not parameters of `calc_spectrum()`, but instead post-calculation convolution steps, so I did not include them as fit parameters. Now `offset` is included. As for `slit`, according to Mr. @minouHub, usually the users know their FTIR spectrometer's stats, and measure `slit` separately, so I have yet implemeted `slit` as fittable parameter.
- Now the users can specify the databank to be fetched into SpectrumFactory with `fetch_databank`, by stating it directly in the script.

The changes can be seen in below script:
```python
from modules.fitting_module import fit_spectrum
from radis import load_spec
import os

# Load experimental spectrum. You can prepare yours, or fetch one of them in the ground-truth folder like below.
os.chdir('/Users/tranhuunhathuy/Documents/GitHub/RADIS-Spectrum-Fitting-Benchmark')
s_experimental = load_spec("./data/LTE/ground-truth/synth-NH3-1-500-2000-cm-1-P10-t1000-v-r-mf0.01-p1-sl1nm.spec")

# Experimental conditions which will be used for spectrum modeling. Basically, these are known ground-truths.
experimental_conditions = {
    "molecule" : "NH3",         # Molecule ID
    "isotope" : "1",            # Isotope ID, can have multiple at once
    "wmin" : 1000,              # Starting wavelength/wavenumber to be cropped out from the original experimental spectrum.
    "wmax" : 1050,              # Ending wavelength/wavenumber for the cropping range.
    "wunit" : "cm-1",           # Accompanying unit of those 2 wavelengths/wavenumbers above.
    "mole_fraction" : 0.01,     # Species mole fraction, from 0 to 1.
    "pressure" : 10,            # Partial pressure of gas, in "bar" unit.
    "path_length" : 1,          # Experimental path length, in "cm" unit.
    "slit" : "1 nm",            # Experimental slit, must be a blank space separating slit amount and unit.
    #"offset" : "-0.2 nm",       # Experimental offset, must be a blank space separating offset amount and unit.
    "databank" : "hitran"       # Databank used for the spectrum calculation. Must be stated.
}

# List of parameters to be fitted.
fit_parameters = {
    "Tgas" : 700,               # Fit parameter, accompanied by its initial value.
    "offset" : "0.1 nm"         # Experimental offset, must be a blank space separating offset amount and unit.
}

# List of bounding ranges applied for those fit parameters above.
bounding_ranges = {
    "Tgas" : [600, 2000],       # Bounding ranges for each fit parameter stated above. You can skip this step, but not recommended.
    "offset" : [-0.3, 0.3]
}

# Fitting pipeline setups.
fit_properties = {
    "method" : "lbfgsb",        # Preferred fitting method from the 17 confirmed methods of LMFIT stated in week 4 blog. By default, "leastsq".
    "fit_var" : "radiance",     # Spectral quantity to be extracted for fitting process, such as "radiance", "absorbance", etc.
    "normalize" : False,        # Either applying normalization on both spectra or not.
    "max_loop" : 150,           # Max number of loops allowed. By default, 100.
    "tol" : 1e-10               # Fitting tolerance, only applicable for "lbfgsb" method.
}


# Conduct the fitting process!
s_best, result, log = fit_spectrum(
    s_exp = s_experimental,
    fit_params = fit_parameters,
    bounds = bounding_ranges,
    model = experimental_conditions,
    pipeline = fit_properties
)


# Now investigate the log

print("\nResidual history: \n")
print(log["residual"])

print("\nFitted values history: \n")
for fit_val in log["fit_vals"]:
    print(fit_val)

print("\nTotal fitting time: ")
print(log["time_fitting"], end = " s\n")
```

And the result seems promising, with `Tgas = 992.043` (ground-truth as 1000) and `offset = -0.1885` (ground-truth as -0.2 nm).

Screen Shot 2022-08-25 at 18 21 24

### CODE REVIEW

**Comment by erwanp** [comment]
Nice ! small comment, can you put the "chdir" outside the main script; because it's specific to your machine; and so that people can just copy/paste your code. 

More important : what's the uncertainty on the temperature you get ?  You say `Tgas = 992.043` ; but it's +/- something. Your fitting algorithm will return a standard deviation . Return `+/- 2 std` to get the usual 95% confidence interval (assuming Gaussian error distribution) 

Note 3 : does the fitting routine already plot the results ? Or do you have to manually add the print() in the user script everytime ?

**Comment by TranHuuNhatHuy** [comment]
Mr. @erwanp , thank you for your comments.
1. The `chdir` won't be included in the official notebooks I generally distribute. As I've just recently switched to Mac OS, I need to do that to make sure it works on my local environment, just my local side.
2. This script is using algorithm `lbfgsb`. As stated by the runtime log:
```
##  Warning: uncertainties could not be estimated:
    this fitting method does not natively calculate uncertainties
    and numdifftools is not installed for lmfit to do this. Use
    `pip install numdifftools` for lmfit to estimate uncertainties
    with this fitting method.
```
So if we use this algorithm, it won't return uncerntainty. This is the most stable and result-looks-best algorithm I've been using till now, so I did not notice this lack of uncertainty...
3. The main function, `fit_spectrum()`, plots the comparative result with `verbose = True` as default (user can shut it down, for sure). It also returns 3 objects:
- `s_best`: the best fit `Spectrum` object.
- `result`: the LMFIT Result object.
- `log`: a `dict` with 3 keys: `residual` for residual history, `fit_vals` for the log of all fit parameters, and `time_fitting` for total fit time.

**Comment by erwanp** [comment]
Install the suggested package to get the uncertainty

**Comment by TranHuuNhatHuy** [comment]
Mr. @erwanp , 

I installed the suggested package `numdifftools` already, but still the uncertainties cannot be determined:
```
##  Warning: uncertainties could not be estimated:
```

I tried switching to other algorithms, or changing the fitting pipeline properties, but the problem persisted. The only time it succeeded, was when I only fit `Tgas` with `"method" : "least_squares"` and `"normalize" : True`. Then, the result was:
```
Succesfully finished the fitting process in 3.090420961380005s.
[[Fit Statistics]]
    # fitting method   = least_squares
    # function evals   = 4
    # data points      = 1
    # variables        = 1
    chi-square         = 5.7058e-09
    reduced chi-square = 5.7058e-09
    Akaike info crit   = -16.9817822
    Bayesian info crit = -18.9817822
[[Variables]]
    Tgas:  993.853759 +/- 1573.45715 (158.32%) (init = 700)
```

Still, quite a strange uncertainty. I searched for this and found some similar problems:
- [Why are uncertainties in parameters sometimes not determined?](https://lmfit.github.io/lmfit-py/faq.html#why-are-uncertainties-in-parameters-sometimes-not-determined)
- [A similar case with someone also using LMFIT for fitting spectra.](https://groups.google.com/g/lmfit-py/c/hIMOTkSvYTg)

Since each fitting algorithm and the pipeline is practically a black box, I believe that for now judging the fit's quality by comparative plot is the only viable way.

**Comment by erwanp** [comment]
There are not any more info on why uncertainties cannot be copouree?

**Comment by TranHuuNhatHuy** [comment]
Mr. @erwanp , after searching extensively, I found out that this problem also occurs a lot in other fitting cases. There are several reasons for this that I collected:

- Some parameters are stuck at initial or boundary values.
- Changes in one or some fitting parameters don't significantly alter the result/residual.
- LMFIT usually cannot calculate uncertainties even with the very same script on different runs.
- Even with `numdifftools` installed, a lot of people still experience this problem.

I believe that this is due to the nature of fitting multiple variables into the spectrum, with some variables like `offset` or `mole_fraction` that only produce insignificant changes between fitting loops, compared to `Tgas` which changes the lineshape significantly. 

Do you have any suggestions, sir?

**Comment by erwanp** [comment]
Reduce the number of fitting parameters. Does it work with Tgas only?
Unfortunately, not having an uncertainty renders the function not usable for scientific purposes, so we have to find a way!

**Comment by TranHuuNhatHuy** [comment]
Yes, it works with Tgas only, but when the number of fitting parameters increases, this phenomenon is very likely to happen.

**Comment by erwanp** [comment]
Fine, it's a good start then. When parameters are independant, you can iterate. 
It may be faster to first find the offset; then the temperature ;rather than explore a 2D-space. 

I you can release an example of your fitting procedure working for Tgas only, for instance compared to the 1T fitting of the Gallery example, it would be very illustrative of your improvements

**Comment by TranHuuNhatHuy** [comment]
Thank you so much Mr. @erwanp for your suggestion. After 2 days investigating the issue, I have several findings and conclusions regarding this phenomenon:

- In some of the LMFIT's examples, their error function returns an array instead of a scalar. This hinted me to try to use `get_diff()` to return the diff array instead of current `get_residual()` to return a diff scalar, hoping that the uncertainties can be determined.
- However, the `get_diff()` function has a horrendous accuracy compared to `get_residual()` function, when being implemented into fitting pipelines, not only my spectrum fitting, but also LMFIT's example ones.
- In addition, the diff scalar actually can return uncertainties in several cases. 

Proof-of-concept scripts will be discussed in our Slack channels.

So, in summary:
- `get_diff()` is questionable, while `get_residual()` is perfect.
- The uncertainties determination has little or nothing to do with error function returning a scalar or an array.
- In the near future, we might need to re-evaluate the `get_diff()` and `get_residual()` functions, as we also spotted some bugs while fitting Mr. Grimaldi's cases.

Now, allow me to merge this update of adding `offset` as fit parameter.
